src/grain/distribution.js

```diff
@@ -0,0 +1,232 @@
+// @flow
+
+/**
+ * In SourceCred, projects regularly distribute Grain to contributors based on
+ * their Cred scores. This is called a "Distribution".
+ *
+ * This module contains the logic for calculating distribution amounts for
+ * contributors. A distribution contains "receipts" showing how much Grain each
+ * contributor will receive, the timestamp of the distribution in question, and the
+ * "strategy" used to distribute grain.
+ *
+ * Currently we support two strategies:
+ * - IMMEDIATE, which distributes a fixed amount of grain based on the cred scores
+ * in the most recent completed time interval
+ * - LIFETIME, which distributes a fixed amount of grain based on cred scores
+ * across
+ * all time, prioritizing paying people who were under-paid historically (i.e.
+ * their lifetime earnings are lower than we would expect given their current
+ * cred score)
+ *
+ * In all cases, the timestamp for the distribution is used to determine which cred
+ * scores are in scope. For example, if you create a immediate distribution with a
+ * timestamp in the past, it will reward people with cred in the past time
+ * period, not the current one.
+ */
+import {sum} from "d3-array";
+import {mapToArray} from "../util/map";
+import {type NodeAddressT} from "../core/graph";
+import {type Grain, multiplyFloat, toFloatRatio, ZERO} from "./grain";
+
+export const DISTRIBUTION_VERSION_1 = 1;
+
+export type DistributionStrategy = ImmediateV1 | LifetimeV1;
+
+export type ImmediateV1 = {|
+  +type: "IMMEDIATE",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type LifetimeV1 = {|
+  +type: "LIFETIME",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type GrainReceipt = {|+address: NodeAddressT, amount: Grain|};
+
+export type DistributionV1 = {|
+  +type: "DISTRIBUTION",
+  +timestampMs: number,
+  +version: number,
+  +strategy: DistributionStrategy,
+  +receipts: $ReadOnlyArray<GrainReceipt>,
+|};
```

Comment:
@hammadj and I discussed this. My position: We already have a mechanism for applying two distribution strategies, which is to model them as two separate distributions that occurred at the same timestamp, and let the ledger logic combine them. Modeling that a single distribution can have multiple sub-distributions adds redundant implementation complexity without any real benefit. Bear in mind that it wouldn't just be a matter of having a single set of receipts and a strategy tuple -- we want to be able to tell that you got X grain from the IMMEDIATE distribution and Y from the LIFETIME distribution, not just that you got (X+Y) from the combined distribution.

Reply:
This is different from the situation in #1697; there we needed extra complexity to keep around information that would otherwise be lost (the mapping between distribution strategies and the minted grain). Here, we can add a bunch of complexity (basically having two different one-to-many abstractions in the ledger space) but we don't get any new information--you could already tell that the two distributions were related because they have the same timestamp.

Reply:
Yeah I agree, you can losslessly represents a single entity using multiple events. However the major difference is, we're essentially creating an [entity](https://en.wikipedia.org/wiki/Domain-driven_design#Building_blocks) of type Distribution, using several imperative mutations. Rather than an atomic immutable declaration.

Which makes it difficult to prove to an observer that these events were indeed created in an atomic way, belong together and are a complete set. For verifiable distributions this can be important.

Let me illustrate how we're mutating the Distribution:
1. Implicitly create the new entity because the first event appears in the event log.
2. Append one strategy's input and output to the entity.
3. Append another strategy's in/output to the entity.

If we needed a good place for Distribution metadata, an explicit creation event would be an option.
1. DistributionCreated {id: 31, title: "..."}
2. DistributionStrategyAdded {id:31, strategy, receipts}
3. DistributionStrategyAdded {id:31, strategy, receipts}

Note that we're never straight up declaring the Distribution as a whole. In the same way as an account, the _current state_ of a Distribution would be the fold of it's events.

So the Distribution acts as a mutable entity like this. What if we don't want it to be mutable? We could create some type of rule to reject new events when the Distribution is frozen.
So something like:

1. DistributionCreated {id: 31, title: "..."}
2. DistributionStrategyAdded {id:31, strategy, receipts}
3. DistributionStrategyAdded {id:31, strategy, receipts}
4. DistributionFrozen {id: 31}

A sneaky attempt to forge an immediate 50k grain for Beanow (:smiling_imp:) weeks after the actual Distribution would be rejected.
5. DistributionStrategyAdded {id:31, ...} (rejected because of DistributionFrozen)

Alternatively, we could define the expected strategies in the `DistributionCreated` event. Or we could declare the whole Distribution in one event and not have mutating event types in the first place.

Reader exercise: how do we prove to an observer that the budgets for strategy 1 and 2 were coordinated from a single configuration (e.g. they stemmed from the same financial policy)?

---

Another difference between the two approaches would be, you don't have fine grained control over how strategies are combined. Your only composition option is to apply sequentially, because that's how event logs are consumed.

For example if we want unused budgets to carry over from the previous one. You can probably add enough data to let that be lossless. And you can get 99% of the way there to proving the events you're looking at were not coincidental or a bug. With a single event that's trivial.

I'm ok with saying v1 doesn't have any special compositions, so we don't support it here. But it's worth keeping in mind.

---

> you could already tell that the two distributions were related because they have the same timestamp.

Because this is _indistinguishable from coincidence_ the timestamp should not be used for this. They should share a unique distribution ID to definitively associate them. As I did with the above illustration.

Reply:
To make this more actionable.

**If we're going with multiple events**

- The events should be named to reflect it's a mutation to a Distribution.
- They need a Distribution ID.
- The [command](https://en.wikipedia.org/wiki/Command%E2%80%93query_separation) which produces the events is responsible for guaranteeing this ID is unique.
- Preferably there is a higher level function which produces an array of events, to be appended to the ledger together.

**For modeling as a single event**

- Would need to preserve the strategy inputs.
- Preserve the strategy composition order.
- Preserve the per-strategy receipts.
- Doesn't strictly needs an ID as the event could be a [value object](https://en.wikipedia.org/wiki/Domain-driven_design#Building_blocks). Referenceable by it's index in the event log (similar to using block height as reference).

The easiest approach is probably to take part of the model you already have and stick it in an array. Letting the v1 imply the way to compose these is by applying sequentially.

```
const event = {
  type,
  timestampMs,
  version,
  strategies: [
    {strategy, receipts},
    {strategy, receipts},
  ],
}
```

Reply:
Holding off on implementing this on this PR since I think there's more to discuss here. Namely, after applying one of the strategies the lifetimeEarnings changes, which could affect the results of the other strategy. e.g. As a new contributor, I get an immediate payment of X grain in a given distribution, and in that same distribution the "balanced" payment sees me as having 0 payout and results in distributing more grain to me than if it knew that I also got an immediate payout.

I think if we are going to reimplement this logic, it deserves its own PR vs just going with "the easiest approach" which might not be the most ideal approach. Will open an issue to keep track of it.